File: concurrency_parallelism/solutions.py
"""Concurrency & Parallelism — Producer-Consumer Queue."""

import queue
import threading
import time
from typing import Callable


def run_producer_consumer(
    num_items: int = 5,
    queue_size: int = 5,
    on_produce: Callable[[str], None] | None = None,
    on_consume: Callable[[str], None] | None = None,
) -> None:
    """Thread-safe producer-consumer using a bounded blocking queue."""
    work_queue: queue.Queue[str] = queue.Queue(maxsize=queue_size)

    def producer() -> None:
        for i in range(num_items):
            item = f"task-{i}"
            work_queue.put(item)
            if on_produce:
                on_produce(item)
            else:
                print(f"[Producer] Produced {item}")
            time.sleep(0.1)

    def consumer() -> None:
        for _ in range(num_items):
            try:
                item = work_queue.get(timeout=2)
                if on_consume:
                    on_consume(item)
                else:
                    print(f"[Consumer] Processed {item}")
                work_queue.task_done()
            except queue.Empty:
                break

    t_producer = threading.Thread(target=producer)
    t_consumer = threading.Thread(target=consumer)
    t_producer.start()
    t_consumer.start()
    t_producer.join()
    t_consumer.join()


if __name__ == "__main__":
    run_producer_consumer()

File: concurrency_parallelism/Solutions.java
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

/** Concurrency & Parallelism — Producer-Consumer Queue. */
public class Solutions {

    /** Thread-safe producer-consumer using a bounded blocking queue. */
    public static void runProducerConsumer(int numItems, int queueSize) throws InterruptedException {
        BlockingQueue<String> queue = new ArrayBlockingQueue<>(queueSize);

        Thread producer = new Thread(() -> {
            try {
                for (int i = 0; i < numItems; i++) {
                    String item = "task-" + i;
                    queue.put(item);
                    System.out.println("[Producer] Produced " + item);
                    Thread.sleep(100);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        Thread consumer = new Thread(() -> {
            try {
                for (int i = 0; i < numItems; i++) {
                    String item = queue.take();
                    System.out.println("[Consumer] Processed " + item);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });

        producer.start();
        consumer.start();
        producer.join();
        consumer.join();
    }

    public static void main(String[] args) throws InterruptedException {
        runProducerConsumer(5, 5);
    }
}
